### PR TITLE
wasm support

### PR DESCRIPTION
**Pull request by AlexErrant**
Do you have any interest in supporting wasm? This PR gets FSRS's `next_interval` working... I think. I haven't called it from Javascript yet - I just ran `wasm-pack build --target web && cat pkg/fsrs.d.ts;`, saw

```ts
export class FSRSwasm {
  free(): void;
/**
* @param {number | undefined} stability
* @param {number} desired_retention
* @param {number} rating
* @returns {number}
*/
  next_interval(stability: number | undefined, desired_retention: number, rating: number): number;
}
```

and declared victory.

### CODE REVIEW

File: src/model.rs [hunk 1]
```diff
@@ -201,6 +202,21 @@ pub struct FSRS<B: Backend = NdArrayBackend> {
     device: B::Device,
 }
 
+#[wasm_bindgen]
+pub struct FSRSwasm(FSRS<NdArrayBackend>);
```
Comment (AlexErrant) [inline src/model.rs:206]: `wasm_bindgen` [can't handle generics](https://github.com/rustwasm/wasm-bindgen/issues/3309), so I made it concrete.

File: src/model.rs [hunk 2]
```diff
@@ -201,6 +202,21 @@ pub struct FSRS<B: Backend = NdArrayBackend> {
...
+impl FSRSwasm {
+    pub fn next_interval(
+        &self,
+        stability: Option<f32>,
+        desired_retention: f32,
+        rating: u32,
+    ) -> u32 {
+        self.0.next_interval(stability, desired_retention, rating)
+    }
+}
```
Comment (AlexErrant) [inline src/model.rs:218]: Using the unfortunately verbose technique described [here](https://users.rust-lang.org/t/how-to-generate-wasm-for-generic-types-for-a-given-specific-type/67164). I am _very_ new to Rust, so there could be something obvious I'm missing. In other functional languages I'd use currying... but not sure if that's possible/reasonable here.

**Comment by dae** [comment]
I'm afraid I don't see the point for such a trivial function that can easily be implemented in JS (and [already is](https://github.com/open-spaced-repetition/fsrs.js/blob/f8d3239d95d20fbea0141e27456a9e0d91047fda/src/index.ts#L312)). You also can't just change the library type, as that would prevent this library from being used in Rust programs - you should make a separate crate instead, which imports this one and exposes what you need.

**Comment by AlexErrant** [comment]
I would like to expose other functions related to training as well, but kept the scope small.

> you should make a separate crate instead

Makes sense! Does this org have any interest in hosting the repo? Seems like The Place for FSRS stuff (but I won't be offended if you want _less_ responsibility).

**Comment by L-M-Sherlock** [comment]
> Makes sense! Does this org have any interest in hosting the repo? Seems like The Place for FSRS stuff (but I won't be offended if you want less responsibility).

I'm glad to host it if you have implemented a useable version. I'm not familiar with wasm, so I don't know whether it's feasible.

By the way, will AnkiWeb support FSRS? @dae

**Comment by dae** [comment]
It should already work, though I haven't integrated the tweaks made in the last few days yet.

**Comment by AlexErrant** [comment]
> It should already work

FYI, maybe an issue with that? https://www.reddit.com/r/Anki/comments/16ucwhu/ankiweb_does_not_use_fsrs4/

Don't feel obligated to respond; I haven't experienced it myself.

**Comment by dae** [comment]
They're using the custom scheduler version, not the built-in version.

**Pull request closed without merging**